File: tests-utils/src/main/java/org/elasticsearch/plugin/readonlyrest/utils/gradle/RorPluginGradleProject.java
package org.elasticsearch.plugin.readonlyrest.utils.gradle;

import org.gradle.tooling.GradleConnector;
import org.gradle.tooling.ProjectConnection;

import java.io.File;
import java.nio.file.Paths;
import java.util.Optional;

public class RorPluginGradleProject {

  private final GradleProperties properties;
  private final File project;
  private final String name;

  public RorPluginGradleProject(String name) {
    this.name = name;
    this.project = esProject(name);
    this.properties = GradleProperties.create(project).orElseThrow(() ->
        new IllegalStateException("cannot load gradle.properties file")
    );
  }

  public static File getRootProject() {
    String projectDir = System.getProperty("project.dir");
    return projectDir != null ? Paths.get(projectDir).toFile() : new File("../");
  }

  public Optional<File> assemble() {
    runTask(name + ":assemble");
    File plugin = new File(project, "build/distributions/" + pluginName());
    if (!plugin.exists()) return Optional.empty();
    return Optional.of(plugin);
  }

  public GradleProperties getProperties() {
    return properties;
  }

  private File esProject(String esProjectName) {
    return new File(getRootProject(), esProjectName);
  }

  private String pluginName() {
    return String.format(
        "%s-%s_es%s.zip",
        properties.getProperty("pluginName"),
        properties.getProperty("pluginVersion"),
        properties.getProperty("esVersion")
    );
  }

  private void runTask(String task) {
    GradleConnector connector = GradleConnector.newConnector().forProjectDirectory(getRootProject());
    ProjectConnection connect = null;
    try {
      connect = connector.connect();
      connect.newBuild().forTasks(task).run();
    } finally {
      if (connect != null) connect.close();
    }
  }
}

File: tests-utils/src/main/scala/tech.beshu.ror.utils/gradle/RorPluginGradleProject.scala
package tech.beshu.ror.utils.gradle

/*
 *    This file is part of ReadonlyREST.
 *
 *    ReadonlyREST is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    ReadonlyREST is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
 */

import java.nio.file.Paths
import better.files._
import java.io.{File => JFile}

import org.gradle.tooling.GradleConnector

import scala.compat.java8.OptionConverters._
import scala.util.Try

object RorPluginGradleProject {
  def fromSystemProperty: RorPluginGradleProject =
    Option(System.getProperty("esModule"))
      .map(new RorPluginGradleProject(_))
      .getOrElse(throw new IllegalStateException("No 'esModule' system property set"))

  def getRootProject: JFile = {
    Option(System.getProperty("project.dir"))
      .map(projectDir => Paths.get(projectDir).toFile)
      .getOrElse(new JFile("."))
  }

  def availableEsModules: List[String] =
    RorPluginGradleProject
      .getRootProject.toScala
      .children
      .filter { f => f.isDirectory }
      .map(_.name)
      .filter(_.matches("^es\\d\\dx$"))
      .toList
}

class RorPluginGradleProject(val name: String) {
  private val project = esProject(name)
  private val esProjectProperties =
    GradleProperties
      .create(project).asScala
      .getOrElse(throw new IllegalStateException("cannot load '" + name + "' project gradle.properties file"))
  private val rootProjectProperties =
    GradleProperties
      .create(RorPluginGradleProject.getRootProject).asScala
      .getOrElse(throw new IllegalStateException("cannot load root project gradle.properties file"))

  def assemble: Option[JFile] = {
    runTask(name + ":ror")
    val plugin = new JFile(project, "build/distributions/" + pluginName)
    if (!plugin.exists) None
    else Some(plugin)
  }

  def getESVersion: String = esProjectProperties.getProperty("esVersion")

  private def esProject(esProjectName: String) = new JFile(RorPluginGradleProject.getRootProject, esProjectName)

  private def pluginName =
    s"${rootProjectProperties.getProperty("pluginName")}-${rootProjectProperties.getProperty("pluginVersion")}_es$getESVersion.zip"

  private def runTask(task: String): Unit = {
    val connector = GradleConnector.newConnector.forProjectDirectory(RorPluginGradleProject.getRootProject)
    val connect = Try(connector.connect())
    val result = connect.map(_.newBuild().forTasks(task).run())
    connect.map(_.close())
    result.fold(throw _, _ => ())
  }
}

